modules/dnn/src/cuda4dnn/csl/memory.hpp

```diff
@@ -0,0 +1,285 @@
+// This file is part of OpenCV project.
+// It is subject to the license terms in the LICENSE file found in the top-level directory
+// of this distribution and at http://opencv.org/license.html.
+
+#ifndef OPENCV_DNN_SRC_CUDA4DNN_CSL_MEMORY_HPP
+#define OPENCV_DNN_SRC_CUDA4DNN_CSL_MEMORY_HPP
+
+#include "error.hpp"
+#include "pointer.hpp"
+
+#include <opencv2/core.hpp>
+
+#include <cuda_runtime_api.h>
+
+#include <cstddef>
+#include <type_traits>
+#include <memory>
+#include <utility>
+
+namespace cv { namespace dnn { namespace cuda4dnn { namespace csl {
+
+    /* @brief smart device pointer with allocation/deallocation methods
+     *
+     * ManagedPtr is a smart shared device pointer which also handles memory allocation.
+     */
+    template <class T>
+    class ManagedPtr {
+        static_assert(!std::is_const<T>::value && !std::is_volatile<T>::value, "T cannot be cv-qualified");
+        static_assert(std::is_standard_layout<T>::value, "T must satisfy StandardLayoutType");
+
+    public:
+        using element_type = T;
+
+        using pointer = DevicePtr<element_type>;
+        using const_pointer = DevicePtr<typename std::add_const<element_type>::type>;
+
+        using size_type = std::size_t;
+
+        ManagedPtr() noexcept : wrapped{ nullptr }, n{ 0 }, capacity{ 0 } { }
+        ManagedPtr(const ManagedPtr&) noexcept = default;
+        ManagedPtr(ManagedPtr&& other) noexcept
+            : wrapped{ std::move(other.wrapped) }, n{ other.n }, capacity { other.capacity }
+        {
+            other.reset();
+        }
+
+        /** allocates device memory for \p count number of element */
+        ManagedPtr(size_type count) {
+            if (count <= 0) {
+                CV_Error(Error::StsBadArg, "number of elements is zero or negative");
+            }
+
+            void* temp = nullptr;
+            CUDA4DNN_CHECK_CUDA(cudaMalloc(&temp, count * sizeof(element_type)));
+
+            auto ptr = typename pointer::pointer(static_cast<element_type*>(temp));
+            wrapped.reset(ptr, [](element_type* ptr) {
+                if (ptr != nullptr) {
+                    /* contract violation for std::shared_ptr if cudaFree throws */
+                    CUDA4DNN_CHECK_CUDA(cudaFree(ptr));
```

Comment:
This deleter can throw if `cudaFree` fails which is mostly UB. It can also lead to termination if the deleter fails while handling an exception.

> Uses the specified deleter d as the deleter. Deleter must be callable for the type T, i.e. d(ptr) must be well formed, have well-defined behavior and not throw any exceptions. 

From [cppreference.com](https://en.cppreference.com/w/cpp/memory/shared_ptr/reset)

Reply:
Indeed.  I recommend just logging the error to stderr and otherwise ignoring it. I think the only way cudaFree() could actually fail is as a result of some other async failure from a previous cuda operation.  That kind of error is going to get reported elsewhere as well so it should be fine.  E.g. if the cuda runtime is bailing out you are going to find out in short order anyway.